src/Bicep.Core/Semantics/ArmTemplateSemanticModel.cs

```diff
@@ -145,97 +144,68 @@ public bool HasErrors()
+                    TemplateParameterType.String => GetPrimitiveType(resolved, t => t.IsTextBasedJTokenType(), LanguageConstants.TypeNameString, LanguageConstants.LooseString),
+                    TemplateParameterType.Int => GetPrimitiveType(resolved, t => t.Type == JTokenType.Integer, LanguageConstants.TypeNameInt, LanguageConstants.LooseInt),
+                    TemplateParameterType.Bool => GetPrimitiveType(resolved, t => t.Type == JTokenType.Boolean, LanguageConstants.TypeNameBool, LanguageConstants.LooseBool),
+                    TemplateParameterType.Array => GetArrayType(resolved),
+                    TemplateParameterType.Object => GetObjectType(SourceFile.Template!, resolved),
+                    TemplateParameterType.SecureString => LanguageConstants.SecureString,
+                    TemplateParameterType.SecureObject => GetObjectType(SourceFile.Template!, resolved, TypeSymbolValidationFlags.IsSecure),
+                    _ => ErrorType.Empty(),
+                };
 
-            return schemaNode.Type.Value switch
-            {
-                TemplateParameterType.String when TryCreateUnboundResourceTypeParameter(GetMetadata(schemaNode), out var resourceType) => resourceType,
-                TemplateParameterType.String => GetPrimitiveType(schemaNode, t => t.IsTextBasedJTokenType(), LanguageConstants.TypeNameString, LanguageConstants.LooseString),
-                TemplateParameterType.Int => GetPrimitiveType(schemaNode, t => t.Type == JTokenType.Integer, LanguageConstants.TypeNameInt, LanguageConstants.LooseInt),
-                TemplateParameterType.Bool => GetPrimitiveType(schemaNode, t => t.Type == JTokenType.Boolean, LanguageConstants.TypeNameBool, LanguageConstants.LooseBool),
-                TemplateParameterType.Array => GetArrayType(schemaNode),
-                TemplateParameterType.Object => GetObjectType(schemaNode),
-                TemplateParameterType.SecureString => LanguageConstants.SecureString,
-                TemplateParameterType.SecureObject => GetObjectType(schemaNode, TypeSymbolValidationFlags.IsSecure),
-                _ => ErrorType.Empty(),
-            };
-        }
+                if (resolved.Nullable?.Value == true)
+                {
+                    bicepType = TypeHelper.CreateTypeUnion(bicepType, LanguageConstants.Null);
+                }
 
-        private TypeSymbol ResolveTypeReference(string reference)
-        {
-            if (!FollowRefsToConcreteTypeDefinition(reference, out var concreteType, out var errorBuilder))
+                return bicepType;
+            }
+            catch (TemplateValidationException tve)
             {
-                return ErrorType.Create(errorBuilder(DiagnosticBuilder.ForDocumentStart()));
+                return ErrorType.Create(DiagnosticBuilder.ForDocumentStart().UnresolvableArmJsonType(tve.TemplateErrorAdditionalInfo.Path ?? "<unknown location>", tve.Message));
```

Comment:
Please could you ensure we've got coverage to assert the error message is correctly formatted? A scenario test or baseline test would probably suffice.

Reply:
There is a test for this in `ArmTemplateSemanticModelTests`, but it didn't include any assertions about the error message or code. Updated that test to cover the message.